### ISSUE TITLE
Wishlist for Backup improvements in v1.21 (mainly for Multi-Tenancy use cases)

### ISSUE THREAD

**Issue by etiennedi** [weaviate/weaviate#3218]
Feel free to leave comments, I'll edit the OP

* combine individual files (to reduce file I/O) -> faster & cheaper
* compress on the fly 
* progress report (e.g. files transferred/total files or MB transferred vs total MB)
* better monitoring (throughput in MB/s, files per second, concurrent transfers, success & error rates)
* creation response should contain a link/href field which can be used to query the status
* status message should contain more metadata
  * progress (see above)
  * `started` and `completed` timestamp
  * link to restore it (right now you need to check the docs or guess what the API is like)

**Comment by rbtz-openai** [comment]
A couple of notes, feel free to ignore:
* if possible, backups should be parallel to be able to utilize IOPS.
* use modern compression (e.g. zstd/lz4): gzip would become perf bottleneck. Configurable compression levels.
* use a library that supports parallel compression to be able to utilize spare CPU.

For APIs, you may consider reusing [google.longrunning.Operation](https://google.aip.dev/151).  That said, that may be a bit of overkill if you are not using the rest of AIPs and [Resource Oriented API](https://google.aip.dev/121) and [Google's API design guidelines in general](https://cloud.google.com/apis/design).

**Comment by redouan-rhazouani** [comment]
I agree that running backups in parallel to leverage IOPS would be advantageous. 
However, achieving optimal performance will depend on how we design the backup process for parallelism. If the outcome of compression is just a few backup files, it may be worthwhile to utilize a compression algorithm that supports parallel compression. However, we should be cautious about the potential impact on node performance, as compression can be CPU-intensive and the node has other responsibilities besides compression. 
To improve performance, one approach could be to decompose the backup into smaller chunks, which can then be compressed and uploaded in parallel. This allows for easy resumption of upload in case of failure. Additionally, we can control the level of concurrency to ensure that the node can efficiently fulfill its other responsibilities.

**Comment by StefanBogdan** [comment]
We have been looking for incremental backups for some time now. This would allow running backups more frequently without having very high costs for the Bucket storage (of curse depends on use case). We had some incident where users deleted the data themselves and wanted to revert the action but it was not possible because there were no backups created.

Also would be very helpful to have an endpoint that lists the created backups so you do not have to record the backup id somewhere else but just list what is available (for a particular  type: `gcs`, `s3`, `filesystem`, or maybe for all types).

**Comment by kcm** [comment]
Incremental backups will be the most performant backups. :) Today, we have to make full backups each time, so it is tempting to focus on making them much faster, but having incremental backups would obviate that need for most users. I have to think that @redouan-rhazouani is correct about balancing resources, if we do end up making large, long-running backups very resource hungry.

Depending on the design, it may also allow for cheap point-in-time restores as well.

**Comment by etiennedi** [comment]
I've created a separate ticket to discuss (and upvote) [Incremental Backups](https://github.com/weaviate/weaviate/issues/3313). The v1.21 version of this ticket is mainly focused on improvements to regular backups.
